MoreLinq/Experimental/SelectAsync.cs

```diff
@@ -0,0 +1,548 @@
+        /// consumed then there's a good chance that some projection work will
+        /// be wasted and a cooperative effort is done that depends on the
+        /// projection function (via a <see cref="CancellationToken"/> as its
+        /// second argument) to cancel those in flight.</para>
+        /// <para>
+        /// The <paramref name="selector"/> function should be designed to be
+        /// thread-agnostic.</para>
+        /// </remarks>
+
+        public static ISelectAsyncEnumerable<TResult> SelectAsync<T, TResult>(
+            this IEnumerable<T> source, Func<T, CancellationToken, Task<TResult>> selector)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (selector == null) throw new ArgumentNullException(nameof(selector));
+
+            return
+                SelectAsyncEnumerable.Create(
+                    options => _(options.MaxConcurrency ?? int.MaxValue,
+                                 options.Scheduler ?? TaskScheduler.Default,
+                                 options.PreserveOrder));
+
+            IEnumerable<TResult> _(int maxConcurrency, TaskScheduler scheduler, bool ordered)
+            {
+                var notices = new BlockingCollection<(Notice, (int, TResult), ExceptionDispatchInfo)>();
+                var cancellationTokenSource = new CancellationTokenSource();
+                var cancellationToken = cancellationTokenSource.Token;
+                var completed = false;
+
+                var enumerator =
+                    source.Index()
+                          .Select(e => (e.Key, Task: selector(e.Value, cancellationToken)))
+                          .GetEnumerator();
+
+                IDisposable disposable = enumerator; // disables AccessToDisposedClosure warnings
+
+                try
+                {
+                    Task.Factory.StartNew(
+                        () =>
+                            CollectToAsync(
+                                enumerator,
+                                e => e.Task,
+                                notices,
+                                (e, r) => (Notice.Result, (e.Key, r), default),
+                                ex => (Notice.Error, default, ExceptionDispatchInfo.Capture(ex)),
+                                (Notice.End, default, default),
+                                maxConcurrency, cancellationTokenSource),
+                        CancellationToken.None,
+                        TaskCreationOptions.DenyChildAttach,
+                        scheduler);
+
+                    var nextKey = 0;
+                    var holds = ordered ? new List<(int, TResult)>() : null;
+
+                    foreach (var (kind, result, error) in notices.GetConsumingEnumerable())
+                    {
+                        if (kind == Notice.Error)
+                            error.Throw();
+
+                        if (kind == Notice.End)
```

Comment:
What is the purpose of this `Notice.End` enum value? AFAICT, if this was missing, we would just the loop one iteration earlier. I feel like I'm missing something